src/de.rs

```diff
@@ -117,965 +70,516 @@ where
             }
-            EitherLifetime::Short(buf) => {
-                let s = Self::convert_str(buf, offset)?;
-                visitor.visit_str(s)
+            types::BadStr(Cow::Owned(buf)) => {
+                visitor.visit_string(unsafe { String::from_utf8_unchecked(buf) })
             }
         }
     }
 
-    fn recursion_checked<F, T>(&mut self, f: F) -> Result<T>
+    #[inline]
+    fn deserialize_string<V>(self, visitor: V) -> Result<V::Value, Self::Error>
     where
-        F: FnOnce(&mut Deserializer<R>) -> Result<T>,
+        V: Visitor<'de>,
     {
-        self.remaining_depth -= 1;
-        if self.remaining_depth == 0 {
-            return Err(self.error(ErrorCode::RecursionLimitExceeded));
-        }
-        let r = f(self);
-        self.remaining_depth += 1;
-        r
+        self.deserialize_str(visitor)
     }
 
-    fn parse_array<V>(&mut self, mut len: usize, visitor: V) -> Result<V::Value>
-    where
-        V: de::Visitor<'de>,
-    {
-        self.recursion_checked(|de| {
-            let value = visitor.visit_seq(SeqAccess { de, len: &mut len })?;
-
-            if len != 0 {
-                Err(de.error(ErrorCode::TrailingData))
-            } else {
-                Ok(value)
-            }
-        })
-    }
-
-    fn parse_indefinite_array<V>(&mut self, visitor: V) -> Result<V::Value>
+    #[inline]
+    fn deserialize_option<V>(self, visitor: V) -> Result<V::Value, Self::Error>
     where
-        V: de::Visitor<'de>,
+        V: Visitor<'de>,
     {
-        self.recursion_checked(|de| {
-            let value = visitor.visit_seq(IndefiniteSeqAccess { de })?;
-            match de.next()? {
-                Some(0xff) => Ok(value),
-                Some(_) => Err(de.error(ErrorCode::TrailingData)),
-                None => Err(de.error(ErrorCode::EofWhileParsingArray)),
-            }
-        })
+        let byte = peek_one(&mut self.reader)?;
+        if byte != marker::NULL && byte != marker::UNDEFINED {
+            let mut de = self.try_step()?;
+            visitor.visit_some(&mut *de)
+        } else {
+            self.reader.advance(1);
+            visitor.visit_none()
+        }
     }
 
-    fn parse_map<V>(&mut self, mut len: usize, visitor: V) -> Result<V::Value>
+    #[inline]
+    fn deserialize_unit<V>(self, visitor: V) -> Result<V::Value, Self::Error>
     where
-        V: de::Visitor<'de>,
+        V: Visitor<'de>,
     {
-        let accept_packed = self.accept_packed;
-        let accept_named = self.accept_named;
-        self.recursion_checked(|de| {
-            let value = visitor.visit_map(MapAccess {
-                de,
-                len: &mut len,
-                accept_named,
-                accept_packed,
-            })?;
-
-            if len != 0 {
-                Err(de.error(ErrorCode::TrailingData))
-            } else {
-                Ok(value)
-            }
-        })
+        let byte = pull_one(&mut self.reader)?;
+        // 0 length array
```

Comment:
is this the only possibility, what about 0 length other things?

Reply:
also, why is an empty array encoded as unit oO

Reply:
That's `cbor4ii` in order to be able to distinguish between `Option::None` and unit. But we serialize it as `NULL`. Good catch, I've changed the code and also added a test for it.